<img src="Assets/yahoo_finance_logo.png" alt="drawing" width="200" />

<h1 style="text-align: center;">stock-market-scraper</h1>

<br/>
<br/>

### `Remember this is for downloading only selected stocks only. `
* **`For mass downloading, head here:`  [stock-market-scraper](https://github.com/Gunjan933/stock-market-scraper/blob/master/stock-market-scraper.py)**

* **`Readme is here:`  [README.md](https://github.com/Gunjan933/stock-market-scraper/blob/master/README.md)**



<br/>
<br/>

# Let's see the scraping idea
<br/>

**Yahoo has gone to a Reactjs front end which means if you analyze the request headers from the client to the backend you can get the actual JSON they use to populate the client side stores.**






<br/>

### Hosts:
* `query1.finance.yahoo.com` HTTP/1.0
* `query2.finance.yahoo.com` HTTP/1.1 [difference between HTTP/1.0 & HTTP/1.1](https://stackoverflow.com/questions/246859/http-1-0-vs-1-1)  

If you plan to use a proxy or persistent connections use `query2.finance.yahoo.com`. But for the purposes of this post the host used for the example URLs is not meant to imply anything about the path it's being used with.

> We will use HTTP/1.1


### Fundamental Data
* `/v10/finance/quoteSummary/AAPL?modules=` (Full list of modules below)

(substitute your symbol for: AAPL)

#### Inputs for the `?modules=` query:

* ```modules = [
   'assetProfile',
   'incomeStatementHistory',
   'incomeStatementHistoryQuarterly',
   'balanceSheetHistory',
   'balanceSheetHistoryQuarterly',
   'cashflowStatementHistory',
   'cashflowStatementHistoryQuarterly',
   'defaultKeyStatistics',
   'financialData',
   'calendarEvents',
   'secFilings',
   'recommendationTrend',
   'upgradeDowngradeHistory',
   'institutionOwnership',
   'fundOwnership',
   'majorDirectHolders',
   'majorHoldersBreakdown',
   'insiderTransactions',
   'insiderHolders',
   'netSharePurchaseActivity',
   'earnings',
   'earningsHistory',
   'earningsTrend',
   'industryTrend',
   'indexTrend',
   'sectorTrend' ]
   ```
#### Example URL:

* `https://query1.finance.yahoo.com/v10/finance/quoteSummary/AAPL?modules=assetProfile%2CearningsHistory`

Querying for: `assetProfile` and `earningsHistory`

The `%2C` is the Hex representation of `,` and needs to be inserted between each module you request. [details about the hex encoding bit](https://stackoverflow.com/questions/6182356/what-is-2c-in-a-url) (if you care)  

<br/>

### Options contracts
* `/v7/finance/options/AAPL` (current expiration)
* `/v7/finance/options/AAPL?date=1579219200` (January 17, 2020 expiration)

#### Example URL:

* `https://query2.yahoo.finance.com/v7/finance/options/AAPL` (current expiration)
* `https://query2.yahoo.finance.com/v7/finance/options/AAPL?date=1579219200` (January 17, 2020 expiration)

Any valid future expiration represented as a UNIX timestamp can be used in the `?date=` query. If you query for the current expiration the JSON response will contain a list of all the valid expirations that can be used in the `?date=` query. ([here is a post explaining converting human readable dates to unix timestamp in Python](https://stackoverflow.com/questions/3682748/converting-unix-timestamp-string-to-readable-date))  

<br/>

### Price
* `/v8/finance/chart/AAPL?symbol=AAPL&period1=0&period2=9999999999&interval=3mo`  

#### Intervals:

* `&interval=3mo` 3 months, going back until initial trading date.
* `&interval=1d` 1 day, going back until initial trading date.
* `&interval=5m` 5 minuets, going back 80(ish) days.
* `&interval=1m` 1 minuet, going back 4-5 days.

How far back you can go with each interval is a little confusing and seems inconsistent. My assumption is that internally yahoo is counting in trading days and my naive approach was not accounting for holidays. Although that's a guess and YMMV.

`period1=`: unix timestamp representation of the date you wish to **start** at. Values below the initial trading date will be rounded up to the initial trading date.

`period2=`: unix timestamp representation of the date you wish to **end** at. Values greater than the last trading date will be rounded down to the most recent timestamp available.

**Note:** *If you query with a `period1=` (start date) that is too far in the past for the interval you've chosen, yahoo will return prices in the `3mo` interval regardless of what interval you requested.*

#### Add pre & post market data

`&includePrePost=true`

#### Add dividends & splits

`&events=div%2Csplit`

#### Example URL:  

* `https://query1.finance.yahoo.com/v8/finance/chart/AAPL?symbol=AAPL&period1=0&period2=9999999999&interval=1d&includePrePost=true&events=div%2Csplit`  

The above request will return all price data for ticker AAPL on a 1 day interval including pre and post market data as well as dividends and splits.

**Note:** *the values used in the price example url for `period1=` & `period2=` are to demonstrate the respective rounding behavior of each input.*


<br/>  

`The above article is taken from `**[here](https://stackoverflow.com/a/47505102/8141330)**`.`

<br/>  


### Dividents and Splits

Yahoo adjusts **all historical** prices to reflect a stock **split**. For example, `ISRG` was trading around $1000 prior to `2017/10/06`. Then on `2017/10/06`, it underwent a 3-for-1 stock split. As you can see, Yahoo's historical prices divided all prices by 3 (both prior to and after `2017/10/06`):
  
    
![](Assets/splits.png)

For **dividends**, let's say stock `ABC` closed at 200 on December 18. Then on December 19, the stock increases in price by `$2` but it pays out a `$1` dividend. In Yahoo's historical prices for XYZ, you will see that it closed at 200 on Dec 18 and 201 on Dec 19. Yahoo factors in the dividend in the **"Adj Close"** column for all the previous days. So the Close for Dec 18 would be 200, but the Adj Close would be 199.

For example, on 2017/09/15, SPY paid out a `$1.235` dividend. Yahoo's historical prices say that SPY's closing price on 2017/09/14 was 250.09, but the Adj Close is 248.85, which is `$1.24` lower. The **Adjusted Close** for the previous days was reduced by the dividend amount.
  
![](Assets/dividents.png)

<br/>  

`The above article is taken from `**[here](https://money.stackexchange.com/a/44146)**`.`

<br/>  



<br/>  

# Now let's get back to some Code to get historic prices of stocks
<br/>  


#### Import some modules:  
* **urllib**: *To get url data*
* **json**: *To handle json files*
* **time**: *To put the program in sleep for some time*
* **os**: *To walk through different directories*
* **difflib**: *To get close matches of strings. Helps to find correct stock from the input user gives.*
* **itertools**: *To repeat same variable to pass in multithreading funciton.*
* **pandas**: *To handle matrix and csv file*
* **datetime**: *To change unix timestamp to normal date and time. Yahoo query uses unix timestamp*

In [1]:
import urllib.request, json , time, os, difflib, itertools
import pandas as pd
from multiprocessing.dummy import Pool
from datetime import datetime

#### Let's make a code snippet which can tell if we have working internet connection or not.
<br/>

In [2]:
try:
    import httplib
except:
    import http.client as httplib

def check_internet():
    conn = httplib.HTTPConnection("www.google.com", timeout=5)
    try:
        conn.request("HEAD", "/")
        conn.close()
        # print("True")
        return True
    except:
        conn.close()
        # print("False")
        return False


<br/>

Now see below, I have opened an arbitrary stock `Igarashi Motors`. In URL can you see the **ticker** for the stock? It is `IGARASHI.BO`

<br/>

![](Assets/chart.png)


<br/>   
<br/>
<br/>

How to get the **ticker**, I will show you later.  

First let us make a **function** that can pull `json data` from yahoo about that stock like below. (I will discuss about the function `parameters` later)
> We will be using query2

<br/>
<br/>
<br/>

![](Assets/query_json.png)

<br/>
<br/>


<br/>

### Now write down the function which will `get_stock_price` for given `query_url`.

#### It will save the stock data as `json` and `csv` inside a folder named "historic_data"
<br/>

In [3]:
def get_historic_price(query_url,json_path,csv_path):
    
    while not check_internet():
        print("Could not connect, trying again in 5 seconds...")
        time.sleep(5)
    
    stock_id=query_url.split("&period")[0].split("symbol=")[1]
    
    if os.path.exists(csv_path+stock_id+'.csv') and os.stat(csv_path+stock_id+'.csv').st_size != 0:
        print("<<<  Historical data of "+stock_id+" already exists, Updating data...")

    try:
        with urllib.request.urlopen(query_url) as url:
            parsed = json.loads(url.read().decode())
    except:
        print("|||  Historical data of "+stock_id+" doesn't exist")
        return
    
    else:
        if os.path.exists(json_path+stock_id+'.json'):
            os.remove(json_path+stock_id+'.json')
        with open(json_path+stock_id+'.json', 'w') as outfile:
            json.dump(parsed, outfile, indent=4)

        try:
            Date=[]
            for i in parsed['chart']['result'][0]['timestamp']:
                Date.append(datetime.utcfromtimestamp(int(i)).strftime('%d-%m-%Y'))

            Low=parsed['chart']['result'][0]['indicators']['quote'][0]['low']
            Open=parsed['chart']['result'][0]['indicators']['quote'][0]['open']
            Volume=parsed['chart']['result'][0]['indicators']['quote'][0]['volume']
            High=parsed['chart']['result'][0]['indicators']['quote'][0]['high']
            Close=parsed['chart']['result'][0]['indicators']['quote'][0]['close']
            Adjusted_Close=parsed['chart']['result'][0]['indicators']['adjclose'][0]['adjclose']

            df=pd.DataFrame(list(zip(Date,Low,Open,Volume,High,Close,Adjusted_Close)),columns =['Date','Low','Open','Volume','High','Close','Adjusted Close'])

            if os.path.exists(csv_path+stock_id+'.csv'):
                os.remove(csv_path+stock_id+'.csv')
            df.to_csv(csv_path+stock_id+'.csv', sep=',', index=None)
            print(">>>  Historical data of "+stock_id+" saved")
            return
        except:
            print(">>>  Historical data of "+stock_id+" exists but has no trading data")

<br/>

#### First we have to set where the `json` and `csv` files will be saved which have been passed to the function `get_historic_price()`
<br/>


In [4]:
json_path = os.getcwd()+os.sep+".."+os.sep+"historic_data"+os.sep+"json"+os.sep
csv_path = os.getcwd()+os.sep+".."+os.sep+"historic_data"+os.sep+"csv"+os.sep

   
#### Then we have to check if these directory exists, if not, then we will use `os.mkdir`
  

In [5]:
if not os.path.isdir(json_path):
    os.makedirs(json_path)
if not os.path.isdir(csv_path):
    os.makedirs(csv_path)


<br/>
<br/>

## Getting tickers

Now as promised I will be showing how to find **historical data**. See below, I have opened historical data of `Igarashi Motors`. Here you can see max time period from which we can pull data for the stock. It stores period as `unix timestamp` in the query.

<br/>

![](Assets/historic_data.png)


<br/>   
<br/>
<br/>

Now let's make the **query**. First set
* `period1 = 0`
* `period2 = 9999999999`
* `interval = 1d`  

See the image below, it's `period1` is greater than `0` and `period2` is lesser than `9999999999`. This produces maximum span period from which data can be pulled.

<br/>
<br/>
<br/>

![](Assets/find_query.png)

<br/>
<br/>


<br/>  

#### Then we need to open our csv file where `yahoo finance tickers` are saved. This is in the `Assets` folder
<br/>

How did I get this? Well here is the **[direct link](http://investexcel.net/wp-content/uploads/2015/01/Yahoo-Ticker-Symbols-September-2017.zip)** to download the **yahoo ticker list (last updated September 2017)**. It would be helpful for the author if you visit **[his website page](http://investexcel.net/all-yahoo-finance-stock-tickers/)**, as his income is through advertisements, and it takes lots of hours to create this type of ticker list.

All right, moving on.

<br/>

## Important Note:

As I will be working on `India`, I will be using a function which gives me the list of stocks which are from India only. If you are from any other country, just change the `country name`, and it will return a list of stocks that are only of `your country`. This shrinking will help us speed up the program. As the original list contains **`106328 stocks`**.


In [6]:
country_name = "india"

#### Let's now make the funciton to shrink the ticker list.

In [7]:
ticker_file_path = "Assets"+os.sep+"Yahoo Ticker Symbols - September 2017.xlsx"
temp_df = pd.read_excel(ticker_file_path)
print("Total stocks:",len(temp_df))
temp_df.head(10)

Total stocks: 106331


Yahoo Stock Tickers                               Unnamed: 1 Unnamed: 2  \
0  http://investexcel.net                                      NaN        NaN   
1                     NaN                                      NaN        NaN   
2                  Ticker                                     Name   Exchange   
3                    OEDV  Osage Exploration and Development, Inc.        PNK   
4                    AAPL                               Apple Inc.        NMS   
5                     BAC              Bank of America Corporation        NYQ   
6                    AMZN                         Amazon.com, Inc.        NMS   
7                       T                                AT&T Inc.        NYQ   
8                    GOOG                            Alphabet Inc.        NMS   
9                      MO                       Altria Group, Inc.        NYQ   

                       Unnamed: 3 Unnamed: 4  Unnamed: 5  Unnamed: 6  \
0                             NaN        NaN         NaN         NaN   
1                             NaN        NaN         NaN         NaN   
2                   Category Name    Country         NaN         NaN   
3                             NaN        USA         NaN         NaN   
4            Electronic Equipment        USA         NaN         NaN   
5              Money Center Banks        USA         NaN         NaN   
6     Catalog & Mail Order Houses        USA         NaN         NaN   
7     Telecom Services - Domestic        USA         NaN         NaN   
8  Internet Information Providers        USA         NaN         NaN   
9                      Cigarettes        USA         NaN         NaN   

                                          Unnamed: 7  
0                                                NaN  
1                                                NaN  
2                                                NaN  
3                                         Samir Khan  
4                     simulationconsultant@gmail.com  
5                                                NaN  
6        This ticker symbol list was downloaded from  
7  http://investexcel.net/all-yahoo-finance-stock...  
8              and was updated on 2nd September 2017  
9                                                NaN

<br/>

#### See the above list is messy, it contains garbage informations. So refining it we get
<br/>

In [8]:
temp_df = temp_df.drop(temp_df.columns[[5, 6, 7]], axis=1)
headers = temp_df.iloc[2]
df  = pd.DataFrame(temp_df.values[3:], columns=headers)
print("Total stocks:",len(df))
df.head(10)

Total stocks: 106328


2 Ticker                                     Name Exchange  \
0   OEDV  Osage Exploration and Development, Inc.      PNK   
1   AAPL                               Apple Inc.      NMS   
2    BAC              Bank of America Corporation      NYQ   
3   AMZN                         Amazon.com, Inc.      NMS   
4      T                                AT&T Inc.      NYQ   
5   GOOG                            Alphabet Inc.      NMS   
6     MO                       Altria Group, Inc.      NYQ   
7    DAL                    Delta Air Lines, Inc.      NYQ   
8     AA                        Alcoa Corporation      NYQ   
9    AXP                 American Express Company      NYQ   

2                   Category Name Country  
0                             NaN     USA  
1            Electronic Equipment     USA  
2              Money Center Banks     USA  
3     Catalog & Mail Order Houses     USA  
4     Telecom Services - Domestic     USA  
5  Internet Information Providers     USA  
6                      Cigarettes     USA  
7                  Major Airlines     USA  
8                        Aluminum     USA  
9                 Credit Services     USA

<br/>  

#### Let's only take the country which is set to `country_name` previously
<br/>

In [9]:
new_df = df[df["Country"].str.lower().str.contains(country_name.lower()) == True]
print("Total stocks:",len(new_df))
new_df.head(10)

Total stocks: 8984


2            Ticker                                        Name Exchange  \
1230  BHARTIARTL.NS                       Bharti Airtel Limited      NSI   
1247    ASHOKLEY.NS                       Ashok Leyland Limited      NSI   
1441  AUROPHARMA.NS                    Aurobindo Pharma Limited      NSI   
1457     AREXMIS.BO                        Arex Industries Ltd.      BSE   
1586    SANWARIA.NS                  Sanwaria Agro Oils Limited      NSI   
1907     ALMONDZ.NS           Almondz Global Securities Limited      NSI   
1962     ADINATH.BO                        Adinath Textiles Ltd      BSE   
2897        SBIN.NS                         State Bank of India      NSI   
3199        BPCL.NS        Bharat Petroleum Corporation Limited      NSI   
3322       MBECL.NS  McNally Bharat Engineering Company Limited      NSI   

2                       Category Name Country  
1230          Wireless Communications   India  
1247       Auto Manufacturers - Major   India  
1441                  Drugs - Generic   India  
1457                              NaN   India  
1586                    Farm Products   India  
1907  Investment Brokerage - National   India  
1962                              NaN   India  
2897               Money Center Banks   India  
3199   Oil & Gas Refining & Marketing   India  
3322              General Contractors   India

<br/>

#### Saving the list of stcks with tickers with `country_name` in a different `csv` file which can be used later.

In [10]:
new_df.to_csv('Assets'+os.sep+country_name+'.csv', sep=',', index=None)

 
#### Now we got our `tickers` which we can find by company names

<br/>

### Important Note
We can scrap historical data for every stock mentioned on yahoo finance. But remember the data size. There are about **`8984`** stocks for only `India`. And if you want for every stocks, there is **`106328`** stocks. Which will take huge resources and time.
Check out my brute force downloader **[stock-market-scraper](stock-market-scraper.py)** for downloading all datasets.

#### List down the company for which the stocks will be downloaded
<br/>

In [11]:
desired_company_list = [ 'Igarashi Motors' ,
                        'State Bank of India' ,
                        'Tata Motors' ,
                        'Tata Consultancy Services' ]

#### Get the ticker list for the companies user entered in the `desired_company_list`
<br/>

In [12]:
ticker_list=[]
for company in desired_company_list:
    try:
        exact_company_name = (difflib.get_close_matches(company, new_df['Name'])[0])
        ticker_for_the_company = new_df.loc[new_df['Name'] == exact_company_name, 'Ticker'].iloc[0]
        ticker_list.append(ticker_for_the_company)
    except:
        print("Company name "+company+" not found.")

<br/>

#### Now create the query urls for the stock `ticker`s. This will bring the query pages, where yahoo finance holds it's historical stock data. 

<br/>

Example query is like this:  `https://query1.finance.yahoo.com/v8/finance/chart/`**ticker**`?symbol=`**ticker**`&period1=0&period2=9999999999&interval=1d&includePrePost=true&events=div%2Csplit`

<br/>  

In [13]:
query_urls=[]
for ticker in ticker_list:
    query_urls.append("https://query1.finance.yahoo.com/v8/finance/chart/"+ticker+"?symbol="+ticker+"&period1=0&period2=9999999999&interval=1d&includePrePost=true&events=div%2Csplit")

### Now get to the stock datas with multithreading.
<br/>

In [14]:
with Pool(processes=len(query_urls)) as pool:
    pool.starmap(get_historic_price, zip(query_urls, itertools.repeat(json_path), itertools.repeat(csv_path)))
print("All downloads completed !")

<<<  Historical data of SBIN.NS already exists, Updating data...
<<<  Historical data of IGARASHI.NS already exists, Updating data...
<<<  Historical data of TATAMOTORS.NS already exists, Updating data...<<<  Historical data of TCS.NS already exists, Updating data...

>>>  Historical data of TCS.NS saved
>>>  Historical data of IGARASHI.NS saved
>>>  Historical data of TATAMOTORS.NS saved
>>>  Historical data of SBIN.NS saved
All downloads completed !


## So like this you can update data everyday by yourself
<br/>

### Source :
* **[Stack Overflow](https://stackoverflow.com/)**
* **[Yahoo Finance](https://in.finance.yahoo.com/)**
* **[Stack Exchange](https://stackexchange.com/)**
* **[Python](https://www.python.org/)**
* **[Geeks for Geeks](https://www.geeksforgeeks.org/)**